src/models/neural_networks/rnn.py

```diff
@@ -0,0 +1,175 @@
+            for i in range(1, len(dense_features))
+        ])
+
+        self.initialize_weights()
+
+    def initialize_weights(self):
+
+        sqrt_k = math.sqrt(1 / self.hidden_size)
+        for parameters in self.rnn.parameters():
+            for pam in parameters:
+                nn.init.uniform_(pam.data, -sqrt_k, sqrt_k)
+
+        for dense_layer in self.dense_layers:
+            nn.init.kaiming_uniform_(dense_layer.weight.data)
+            nn.init.constant_(dense_layer.bias.data, 0)
+
+    def forward(self, x, pred_month=None):
+
+        sequence_length = x.shape[1]
+
+        hidden_state = torch.zeros(1, x.shape[0], self.hidden_size)
+        cell_state = torch.zeros(1, x.shape[0], self.hidden_size)
+
+        if x.is_cuda:
+            hidden_state = hidden_state.cuda()
+            cell_state = cell_state.cuda()
+
+        for i in range(sequence_length):
+            # The reason the RNN is unrolled here is to apply dropout to each timestep;
+            # The rnn_dropout argument only applies it after each layer.
+            # https://www.tensorflow.org/api_docs/python/tf/nn/rnn_cell/DropoutWrapper
+            input_x = x[:, i, :].unsqueeze(1)
+            _, (hidden_state, cell_state) = self.rnn(input_x,
+                                                     (hidden_state, cell_state))
+            hidden_state = self.dropout(hidden_state)
+
+        x = hidden_state.squeeze(0)
+
+        if self.include_pred_month:
+            x = torch.cat((x, pred_month), dim=-1)
+        for layer_number, dense_layer in enumerate(self.dense_layers):
+            x = dense_layer(x)
+        return x
+
+
+class UnrolledRNN(nn.Module):
+    """An unrolled RNN. The motivation for this is mainly so that we can explain this model using
+    the shap deep explainer, but also because we unroll the RNN anyway to apply dropout.
+    """
+
+    def __init__(self, input_size, hidden_size, batch_first=True):
+        super().__init__()
+
+        self.input_size = input_size
+        self.hidden_size = hidden_size
+        self.batch_first = batch_first
+
+        self.forget_gate = nn.Sequential(*[
+            nn.Linear(in_features=input_size + hidden_size, out_features=hidden_size,
+                      bias=True), nn.Sigmoid()])
+
+        self.update_gate = nn.Sequential(*[
+            nn.Linear(in_features=input_size + hidden_size, out_features=hidden_size,
+                      bias=True), nn.Sigmoid()
+        ])
+
+        self.update_candidates = nn.Sequential(*[
+            nn.Linear(in_features=input_size + hidden_size, out_features=hidden_size,
+                      bias=True), nn.Tanh()
+        ])
+
+        self.output_gate = nn.Sequential(*[
+            nn.Linear(in_features=input_size + hidden_size, out_features=hidden_size,
+                      bias=True), nn.Sigmoid()
+        ])
+
+        self.cell_state_activation = nn.Tanh()
+
+    def forward(self, x, state):
+        hidden, cell = state
```

Comment:
what is the type of `state`? Is `state` the current information flowing through the forget/update gates?

Reply:
![Screenshot 2019-07-16 at 10 02 31](https://user-images.githubusercontent.com/21049064/61281044-da4be380-a7b0-11e9-9d32-e36de0dd1715.png)

Is it the `C` here?
From http://colah.github.io/posts/2015-08-Understanding-LSTMs/

Reply:
State is a tuple `(hidden, cell)` (see line 157) - so its `(h_{t-1}, C_{t-1})`